src/liballoc_jemalloc/lib.rs

```diff
@@ -201,27 +201,47 @@ mod contents {
                                             align: usize,
                                             excess: *mut usize,
                                             err: *mut u8) -> *mut u8 {
-        let p = __rde_alloc(size, align, err);
-        if !p.is_null() {
-            *excess = size;
+        let flags = align_to_flags(align);
+        let ptr = mallocx(size as size_t, flags) as *mut u8;
+        if ptr.is_null() {
+            let layout = Layout::from_size_align_unchecked(size, align);
+            ptr::write(err as *mut AllocErr,
+                       AllocErr::Exhausted { request: layout });
+        } else {
+            *excess = nallocx(size as size_t, flags) as usize;
```

Comment:
I did it like this at first in #45512 but I think this is wrong. `nallocx` returns the total allocation size, so `excess` should be that - size.

Reply:
My understanding is that `excess` (despite the name, but that's probably to ) should have the entire capacity, not the extra.

```
    unsafe fn realloc_excess(&mut self,
                             ptr: *mut u8,
                             layout: Layout,
                             new_layout: Layout) -> Result<Excess, AllocErr> {
        let mut err = ManuallyDrop::new(mem::uninitialized::<AllocErr>());
        let mut size = 0;
        let ptr = __rust_realloc_excess(ptr,
                                        layout.size(),
                                        layout.align(),
                                        new_layout.size(),
                                        new_layout.align(),
                                        &mut size,
                                        &mut *err as *mut AllocErr as *mut u8);
        if ptr.is_null() {
            Err(ManuallyDrop::into_inner(err))
        } else {
            Ok(Excess(ptr, size))
        }
    }
```

Reply:
I just commented on the #32838 tracking issue about this, I found it really confusing. Not only in the Allocator trait, but also because it propagates back to the wrappers over the allocator functions like here.

Reply:
Yes, the classic naming problem.